include/bactria/Plugin.hpp

```diff
@@ -0,0 +1,170 @@
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ *  Licence permissions and limitations under the Licence.
+ */
+
+#ifndef BACTRIA_PLUGIN_HPP
+#   define BACTRIA_PLUGIN_HPP
+
+#   include <bactria/detail/PluginInterface.hpp>
+#   include <bactria/detail/POSIX.hpp>
+#   include <bactria/detail/Win32.hpp>
+
+namespace bactria
+{
+    namespace plugin
+    {
+        using plugin_handle_t = detail::plugin_handle_t;
+
+        [[nodiscard]] auto load_plugin() -> plugin_handle_t
+        {
+            auto const path = std::getenv("BACTRIA_PLUGIN");
+            if(path != nullptr)
+            {
+                auto handle = detail::open_plugin(path);
+
+                detail::load_func(handle, detail::create_event_ptr, "bactria_plugin_create_event");
+                detail::load_func(handle, detail::destroy_event_ptr, "bactria_plugin_destroy_event");
+                detail::load_func(handle, detail::fire_event_ptr, "bactria_plugin_fire_event");
+
+                detail::load_func(handle, detail::create_range_ptr, "bactria_plugin_create_range");
+                detail::load_func(handle, detail::destroy_range_ptr, "bactria_plugin_destroy_range");
+                detail::load_func(handle, detail::start_range_ptr, "bactria_plugin_start_range");
+                detail::load_func(handle, detail::stop_range_ptr, "bactria_plugin_stop_range");
+
+                detail::load_func(handle, detail::create_sector_ptr, "bactria_plugin_create_sector");
+                detail::load_func(handle, detail::destroy_sector_ptr, "bactria_plugin_destroy_sector");
+                detail::load_func(handle, detail::enter_sector_ptr, "bactria_plugin_enter_sector");
+                detail::load_func(handle, detail::leave_sector_ptr, "bactria_plugin_leave_sector");
+                detail::load_func(handle, detail::sector_summary_ptr, "bactria_plugin_sector_summary");
+
+                detail::load_func(handle, detail::create_phase_ptr, "bactria_plugin_create_phase");
+                detail::load_func(handle, detail::destroy_phase_ptr, "bactria_plugin_destroy_phase");
+                detail::load_func(handle, detail::enter_phase_ptr, "bactria_plugin_enter_phase");
+                detail::load_func(handle, detail::leave_phase_ptr, "bactria_plugin_leave_phase");
+
+                return handle;
+            }
+
+            // getenv failed
+            throw std::runtime_error{"BACTRIA_PLUGIN not specified"};
+        }
+
+        auto unload_plugin(plugin_handle_t handle) noexcept
+        {
+            detail::close_plugin(handle);
+        }
+
+        [[nodiscard, gnu::always_inline, clang::acquire_handle("bactria event")]]
+        inline auto create_event(const char* name, std::uint32_t color, std::uint32_t category) noexcept
+        {
+            return (detail::create_event_ptr)(name, color, category);
+        }
+
+        [[gnu::always_inline]]
```

Comment:
icc and MSVC disapprove :)

Reply:
They should just generate a warning, right? Warnings can be turned off, I just need to look up the flags once I get to test bactria with these compilers.

Reply:
Sure, don't worry about them now! ;)